File: 2047-find-a-peak-element-ii/find-a-peak-element-ii.java
class Solution {
    public int[] findPeakGrid(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;

        int l = 0 , r = m-1;

        while(l<=r){
            int mid = l + ((r-l)>>1);
            int mxrow = 0;
            int mxm = mat[mxrow][mid];
            for(int i=1;i<n;i++){
                if(mxm<mat[i][mid]){
                    mxm = mat[i][mid];
                    mxrow = i;
                }
            }

            if(mid>0 && mat[mxrow][mid-1]>mxm) r = mid-1;
            else if(mid<m-1 && mat[mxrow][mid+1]> mxm) l = mid+1;
            else return new int[]{mxrow , mid};
        }
        return new int[]{-1,-1};
    }
}

File: 2047-find-a-peak-element-ii/find-a-peak-element-ii.cpp
class Solution {
public:
    vector<int> findPeakGrid(vector<vector<int>>& mat) {
        int m=mat.size(), n=mat[0].size();
        int l=0, r=n-1;

        while(l<=r){
            int mid=l+(r-l)/2;
            int row=0; 
            for(int i=0; i<m; i++){
                if(mat[row][mid]<mat[i][mid]) row=i;
            }
            int val=mat[row][mid];
            if((mid==0 || val>mat[row][mid-1]) && (mid==n-1 || val>mat[row][mid+1])) return {row, mid};
            else if(mid>0&&val<mat[row][mid-1]) r=mid-1;
            else l=mid+1;
        }
        return {-1, -1};
    }
};
